p4-16/spec/P4-16-spec.mdk

```diff
@@ -2717,13 +2767,64 @@ X.v1  // reference to v1
 v1    // error - v1 is not in the top-level namespace
 ~ End P4Example
 
-Similar to errors, `enum` expressions only support equality (`==`)
-and inequality (`!=`) comparisons. Expressions whose type is an `enum`
+Similar to errors, `enum` expressions without a specified backing type only support equality (`==`)
+and inequality (`!=`) comparisons. Expressions whose type is an `enum` without a specified backing type
 cannot be cast to or from any other type.
 
-Note that if an `enum` value appears in the control-plane API, the
+An `enum` that specifies a backing type, such as the following:
+
+~ Begin P4Example
+enum bit<8> E {
+  e1 = 0,
+  e2 = 1,
+  e3 = 2
+}
+~ End P4Example
+
+also supports explicit casts to and from the backing type.  For instance, the
+following code:
+
+~ Begin P4Example
+bit<8> x;
+E e1 = E.e2;
+E e2;
+
+x  = (bit<8>) e1; // sets x to 1
+e2 = (E) x;       // sets e to E.e2
+~ End P4Example
+
+casts `e1`, which was initialized to `E.e2` to a `bit<8>`, using the specified
+numeric representation for `E.e2`, `1`.  The variable `e2` is then set to the
+symbolic value `E.e2`, which corresponds to the numeric value `1`.
+
+Note that while it is always safe to cast from an `enum` to its numeric type,
+it may not be th case that every numeric value specified by the backing type
+has a related symbol in the `enum`.  For instance, the following code:
+
+~ Begin P4Example
+bit<8> x = 5;
+E e = (E) x; // sets e to an unspecified value
+~ End P4Example
+
+sets `e` to an unknown value, since there is no symbol corresponding to the
+numeric value `5`.  While the value of `e` is unspecified, it must not
+correspond to any symbol defined within enum `E`.  This is because casting from
+a numeric type to an `enum` type is required obey the following semantics:
+
+Given an enum type $E$ with underlying type `bit<W>` and the set of mappings
+of $(s, n)$ where $S$ is the set of symbols ($s$) and $N$ is the set of numeric
+representations ($n$), the following properties must hold:
+
+1. if an integer $x \notin N$ then $\forall e \in S$, `(E)x != e`
+1. $\forall x \in [0, 2^W]$, `(bit<W>)(E)x == x`
+1. $\forall e \in S$, `(E)(bit<W>)e == e`
```

Comment:
So if I roughly translate this into English, if an enum type with a backing type `bit<W>`, where the numeric values defined by the enum definition do _not_ include all possible `2^W` values of the backing type, then it is guaranteed that variables of that enum type might not be equal to any of the named values in the enum type?

```
enum bit<2> MyPartialEnum_t {
    VALUE_A = 2w0,
    VALUE_B = 2w1,
    VALUE_C = 2w2
}

MyPartialEnum_t x;

if (x == MyPartialEnum_t.VALUE_A) {
    // some code here
} else if (x == MyPartialEnum_t.VALUE_B) {
    // some code here
} else if (x == MyPartialEnum_t.VALUE_C) {
    // some code here
} else {
    // A P4 compiler MUST ASSUME that this branch can be executed
    // some code here
}
```

Reply:
Should one (or both) of these examples be added as explanation to the math?

Reply:
It's not obvious that the first example is unique to enums with backing types. If uninitialized enum variables are intended to be "undefined" only within valid entries of the enum, it might be good to state that explicitly.

However, if that's the case, (compiler) implementations will already need to initialize non-backed enum variables in most cases, so why not require the same for enums with backing types?

Reply:
@jamescoole-cisco I actually wrote a change to the P4_16 language spec, committed after v1.0.0 was published, that says something about exactly this case, but had forgotten I wrote that part.  See https://github.com/p4lang/p4-spec/issues/611#issuecomment-386702605 for details.  Unless that text is changed further in the spec, P4_16 compilers are already free to make uninitialized enum and error values unequal to any of their named members.

Reply:
Given my previous comment, I am not against having example code like these in the spec, but if so, they should be clear that the examples apply for enum's with _or without_ underlying types.

I think having no code examples would also be fine, but adding a sentence that mentions that _all_ values with type enum can be not-equal to any of their named members, with a cross-reference to the section "## Operations on headers { #sec-ops-on-hdrs }" (Madoko syntax copied from the .mdk file)

Reply:
I agree it would be good to copy or move the discussion of how undefined both types of enums can be into section, since it applies in situations besides header fields. But IDK if the example is necessary to get the point across.